### DISCUSSION TITLE
TheiaCloud as GitPod

### DISCUSSION THREAD

**Discussion by we-wake-act21** [eclipse-theia/theia-cloud#317] [Q&A]
Hi,

I am working on implementing a service similar to GitPod, where a Git repository can be specified as a `URL` query parameter, and the repository is then cloned and set up as a workspace in a pod.

I am considering two approaches to achieve this:

1. Use the existing APIs in Theia Cloud to create a pod and initialize the pod with a `git clone` command.
2. Add an extension to Theia that triggers whenever a query parameter containing the Git repository URL is passed, which then handles the cloning.

Which approach would be more suitable for achieving this functionality, or is there another approach I could consider?

**Comment by jfaltermeier** [comment]
Hi,

We previously prototyped a similar feature in this PR: https://github.com/eclipsesource/theia-cloud/pull/182. Currently, we lack the resources to continue developing this feature ourselves, but you can find more information about our support options here: https://theia-cloud.io/support/.

In that PR, we introduced a concept of init operations for sessions, which have an ID and a list of arguments. The initial implementation included a git init operation that utilized an init-container within the pod to handle cloning, assuming that clone secrets were available as Kubernetes secrets. The advantage of this approach is its broad applicability across different applications, as the cloning process is handled by the init-container, thereby requiring no changes from the users' applications.

This method aligns with your first option, and I believe it offers a robust long-term solution.

Alternatively, your second option allows for implementation without modifying Theia Cloud. You could pass cloning parameters as environment variables through the session and handle the cloning in the Theia extension or your Docker startup script.
